File: app/src/main/java/com/aar/app/wordsearch/features/gamehistory/GameHistoryActivity.java
package com.aar.app.wordsearch.features.gamehistory;

import android.arch.lifecycle.ViewModelProviders;
import android.content.Intent;
import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.TextView;

import com.aar.app.wordsearch.R;
import com.aar.app.wordsearch.WordSearchApp;
import com.aar.app.wordsearch.easyadapter.MultiTypeAdapter;
import com.aar.app.wordsearch.features.FullscreenActivity;
import com.aar.app.wordsearch.features.ViewModelFactory;
import com.aar.app.wordsearch.features.gameplay.GamePlayActivity;
import com.aar.app.wordsearch.model.GameDataInfo;

import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class GameHistoryActivity extends FullscreenActivity {

    @BindView(R.id.textEmpty)
    TextView mTextEmpty;
    @BindView(R.id.recyclerView)
    RecyclerView mRecyclerView;

    @Inject
    ViewModelFactory mViewModelFactory;
    private GameHistoryViewModel mViewModel;

    private MultiTypeAdapter mAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_game_history);

        ((WordSearchApp) getApplication()).getAppComponent().inject(this);
        ButterKnife.bind(this);

        initRecyclerView();

        mViewModel = ViewModelProviders.of(this, mViewModelFactory).get(GameHistoryViewModel.class);
        mViewModel.getOnGameDataInfoLoaded().observe(this, this::onGameDataInfoLoaded);
    }

    @Override
    protected void onResume() {
        super.onResume();
        mViewModel.loadGameHistory();
    }

    @OnClick(R.id.btnClear)
    public void onButtonClearClick() {
        mViewModel.clear();
    }

    private void onGameDataInfoLoaded(List<GameDataInfo> gameDataInfos) {
        if (gameDataInfos.isEmpty()) {
            mRecyclerView.setVisibility(View.GONE);
            mTextEmpty.setVisibility(View.VISIBLE);
        } else {
            mRecyclerView.setVisibility(View.VISIBLE);
            mTextEmpty.setVisibility(View.GONE);
            mAdapter.setItems(gameDataInfos);
        }
    }

    private void initRecyclerView() {
        GameDataAdapterDelegate gameDataAdapterDelegate = new GameDataAdapterDelegate();
        gameDataAdapterDelegate.setOnClickListener(new GameDataAdapterDelegate.OnClickListener() {
            @Override
            public void onClick(GameDataInfo gameDataInfo) {
                Intent intent = new Intent(GameHistoryActivity.this, GamePlayActivity.class);
                intent.putExtra(GamePlayActivity.EXTRA_GAME_ROUND_ID, gameDataInfo.getId());
                startActivity(intent);
            }

            @Override
            public void onDeleteClick(GameDataInfo gameDataInfo) {
                mViewModel.deleteGameData(gameDataInfo);
            }
        });

        mAdapter = new MultiTypeAdapter();
        mAdapter.addDelegate(gameDataAdapterDelegate);

        mRecyclerView.setAdapter(mAdapter);
        mRecyclerView.setLayoutManager(new LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false));
    }
}

File: app/src/main/java/ke/choxxy/wordsearch/gamehistory/GameHistoryActivity.kt
package ke.choxxy.wordsearch.gamehistory

import android.content.Intent
import android.os.Bundle
import android.view.View
import androidx.activity.viewModels
import androidx.recyclerview.widget.LinearLayoutManager
import butterknife.ButterKnife
import ke.choxxy.wordsearch.FullscreenActivity
import ke.choxxy.wordsearch.R
import ke.choxxy.wordsearch.commons.viewBinding
import ke.choxxy.wordsearch.databinding.ActivityGameHistoryBinding
import ke.choxxy.wordsearch.easyadapter.MultiTypeAdapter
import ke.choxxy.wordsearch.gameplay.GamePlayActivity
import ke.choxxy.wordsearch.model.GameDataInfo

class GameHistoryActivity : FullscreenActivity() {

    private val binding by viewBinding(ActivityGameHistoryBinding::inflate)
    private val mViewModel: GameHistoryViewModel by viewModels()
    private lateinit var mAdapter: MultiTypeAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_game_history)
        ButterKnife.bind(this)
        initRecyclerView()
        mViewModel.onGameDataInfoLoaded.observe(this) { gameDataInfos: List<GameDataInfo> ->
            onGameDataInfoLoaded(
                gameDataInfos
            )
        }

        binding.btnClear.setOnClickListener {
            mViewModel.clear()
        }
    }

    override fun onResume() {
        super.onResume()
        mViewModel.loadGameHistory()
    }

    private fun onGameDataInfoLoaded(gameDataInfos: List<GameDataInfo>) {
        if (gameDataInfos.isEmpty()) {
            binding.recyclerView.visibility = View.GONE
            binding.textEmpty.visibility = View.VISIBLE
        } else {
            binding.recyclerView.visibility = View.VISIBLE
            binding.textEmpty.visibility = View.GONE
            mAdapter.setItems(gameDataInfos)
        }
    }

    private fun initRecyclerView() {
        val gameDataAdapterDelegate = GameDataAdapterDelegate()
        gameDataAdapterDelegate.setOnClickListener(object :
            GameDataAdapterDelegate.OnClickListener {
            override fun onClick(gameDataInfo: GameDataInfo) {
                val intent = Intent(this@GameHistoryActivity, GamePlayActivity::class.java)
                intent.putExtra(GamePlayActivity.EXTRA_GAME_ROUND_ID, gameDataInfo.id)
                startActivity(intent)
            }

            override fun onDeleteClick(gameDataInfo: GameDataInfo) {
                    mViewModel.deleteGameData(gameDataInfo)
            }
        })
        mAdapter = MultiTypeAdapter()
        mAdapter.addDelegate(gameDataAdapterDelegate)
        binding.recyclerView.adapter = mAdapter
        binding.recyclerView.layoutManager =
            LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false)
    }
}